api/etcd/list.go

```diff
@@ -4,200 +4,61 @@ import (
-			rangeEnd = clientv3.GetPrefixRangeEnd(s.rulesPath(prefix, ""))
-		} else {
-			rangeEnd = clientv3.GetPrefixRangeEnd(s.latestVersionPath(prefix))
-		}
-		options = append(options, clientv3.WithRange(rangeEnd))
+		opts = append(opts, clientv3.WithRange(clientv3.GetPrefixRangeEnd(s.signaturesPath(""))))
 	} else {
-		key = prefix
-		options = append(options, clientv3.WithPrefix())
+		key = s.signaturesPath("")
+		opts = append(opts, clientv3.WithPrefix())
 	}
 
-	options = append(options, clientv3.WithLimit(int64(limit)))
-
-	switch {
-	case opt.PathsOnly:
-		return s.listPathsOnly(ctx, key, prefix, limit, options)
-	case opt.AllVersions:
-		return s.listAllVersions(ctx, key, prefix, limit, options)
-	default:
-		return s.listLastVersion(ctx, key, prefix, limit, options)
-	}
-}
+	// limit the number of results
+	opts = append(opts, clientv3.WithLimit(int64(opt.GetLimit())))
 
-// listPathsOnly returns only the path for each ruleset.
-func (s *RulesetService) listPathsOnly(ctx context.Context, key, prefix string, limit int, opts []clientv3.OpOption) (*api.Rulesets, error) {
-	opts = append(opts, clientv3.WithKeysOnly())
-	resp, err := s.Client.KV.Get(ctx, s.latestVersionPath(key), opts...)
+	// fetch signatures, the paths will be computed from signature keys
+	resp, err := s.Client.KV.Get(ctx, key, opts...)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to fetch versions")
-	}
-
-	// if a prefix is provided it must always return results
-	// otherwise it doesn't exist.
-	if resp.Count == 0 && prefix != "" {
-		return nil, api.ErrRulesetNotFound
+		return nil, errors.Wrap(err, "failed to fetch signatures")
 	}
 
 	rulesets := api.Rulesets{
-		Rulesets: make([]regula.Ruleset, len(resp.Kvs)),
-	}
-	rulesets.Revision = strconv.FormatInt(resp.Header.Revision, 10)
-	for i, pair := range resp.Kvs {
-		p := strings.TrimPrefix(string(pair.Key), s.latestVersionPath("")+"/")
-		rulesets.Rulesets[i].Path = p
-	}
-
-	if len(rulesets.Rulesets) < limit || !resp.More {
-		return &rulesets, nil
+		Revision: strconv.FormatInt(resp.Header.Revision, 10),
 	}
 
-	lastEntry := rulesets.Rulesets[len(rulesets.Rulesets)-1]
-
-	// we want to start immediately after the last key
-	rulesets.Continue = base64.URLEncoding.EncodeToString([]byte(lastEntry.Path + "\x00"))
-
-	return &rulesets, nil
-}
-
-// listLastVersion returns only the latest version for each ruleset.
-func (s *RulesetService) listLastVersion(ctx context.Context, key, prefix string, limit int, opts []clientv3.OpOption) (*api.Rulesets, error) {
-	resp, err := s.Client.KV.Get(ctx, s.latestVersionPath(key), opts...)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to fetch latest keys")
-	}
-
-	// if a prefix is provided it must always return results
-	// otherwise it doesn't exist.
-	if resp.Count == 0 && prefix != "" {
-		return nil, api.ErrRulesetNotFound
-	}
-
-	ops := make([]clientv3.Op, 0, resp.Count)
-	txn := s.Client.KV.Txn(ctx)
+	rulesets.Paths = make([]string, 0, len(resp.Kvs))
```

Comment:
Assuming resp.Kvs is a map, it might be worth considering sorting the final slice so that the results are deterministic?
:tulip: :tulip:

Reply:
Actually, `resp.Kvs` is a slice so the order is kept (I know the name is not the best)